src/main/java/fi/aalto/cs/apluscourses/presentation/filter/Filter.java

```diff
@@ -0,0 +1,17 @@
+package fi.aalto.cs.apluscourses.presentation.filter;
+
+import java.util.Optional;
+import org.jetbrains.annotations.NotNull;
+
+public interface Filter {
```

Comment:
I'm a bit lost here...

![Screenshot from 2020-09-15 18:34:57](https://user-images.githubusercontent.com/9991098/93232155-3d088880-f782-11ea-8925-e157b18d488b.png)

Reply:
`Filter` is like Java's `Predicate<T>` (with the exception that it returns true/false/empty as discussed elsewhere).

`TypedFilter<T>` is an abstract class that can be extended to implement such filters that return empty if the item is not type of `T` and otherwise true/false based on logic written into subclass.

`Option` is what we use for these kind of filtering options: "show non-submittable exercises", "show exercises with submission count exceeded" etc. They are given as a constructor parameter a filter that defines those target objects (e.g. for "show non-submittable exercises" option is given such a filter that returns true for non-submittable exercises, false for other exercises and empty for non-exercises). Option has property `isSelected`. If it's true, it always returns empty (that is, it doesn't make any restrictions to what should be shown). If `isSelected` is false, it returns false for its target objects and otherwise true/empty. E.g. when "show non-submittable exercises" option is not selected, it hides non-submittable exercises (by returning false for them).

`LogicalFilter` is just a class I wrote to be base for "and-filter" (`LogicalFilter.Conjunction`) and "or-filter" (turned out there was no need for it), that is filters that compose multiple filters and applies a logical operator (AND/OR) to them. As there's only one such (AND i.e. conjunction) currently in use, the whole class could be replace with more specific `AndFilter` class (which could also has proper short-circuiting rules to improve performance).

`Options` is a collection of `Option` objects and applies AND-logic to them (thus inheriting `LogicalFilter.Conjunction`). An `Options` object is what we apply to the exercise tree and it makes sure that each of its `Option`s is applied to the tree elements.

Reply:
Replacing `LogicalFilter` with the more concrete `AndFilter` has my strong support. Our code already has plenty of base classes which aren't really necessary :smile: